zqd/space/space.go

```diff
@@ -32,232 +29,180 @@ var (
-		PcapPath:    s.PcapPath(),
-		PcapSize:    pcapsize,
-	}
-	return spaceInfo, nil
+	g.deletePending = true
+	g.opMutex.Unlock()
+
+	close(g.cancelChan)
+	g.wg.Wait()
+	return nil
 }
 
-// PcapSearch returns a *pcap.SearchReader that streams all the packets meeting
-// the provided search request. If pcaps are not supported in this Space,
-// ErrPcapOpsNotSupported is returned.
-func (s *Space) PcapSearch(ctx context.Context, req api.PcapSearch) (*SearchReadCloser, error) {
-	if s.PcapPath() == "" || !s.hasFile(PcapIndexFile) {
-		return nil, ErrPcapOpsNotSupported
-	}
-	index, err := pcap.LoadIndex(s.DataPath(PcapIndexFile))
-	if err != nil {
-		return nil, err
-	}
-	var search *pcap.Search
-	switch req.Proto {
-	case "tcp":
-		flow := pcap.NewFlow(req.SrcHost, int(req.SrcPort), req.DstHost, int(req.DstPort))
-		search = pcap.NewTCPSearch(req.Span, flow)
-	case "udp":
-		flow := pcap.NewFlow(req.SrcHost, int(req.SrcPort), req.DstHost, int(req.DstPort))
-		search = pcap.NewUDPSearch(req.Span, flow)
-	case "icmp":
-		search = pcap.NewICMPSearch(req.Span, req.SrcHost, req.DstHost)
-	default:
-		return nil, fmt.Errorf("unsupported proto type: %s", req.Proto)
-	}
-	f, err := fs.Open(s.PcapPath())
-	if err != nil {
-		return nil, err
-	}
-	slicer, err := pcap.NewSlicer(f, index, req.Span)
-	if err != nil {
-		f.Close()
-		return nil, err
-	}
-	pcapReader, err := pcapio.NewReader(slicer)
-	if err != nil {
-		f.Close()
-		return nil, err
-	}
-	r, err := search.Reader(ctx, pcapReader)
-	if err != nil {
-		f.Close()
-		return nil, err
+func loadSpaces(path string, conf config) ([]Space, error) {
+	datapath := conf.DataPath
+	if datapath == "." {
+		datapath = path
 	}
-	return &SearchReadCloser{r, f}, nil
 
-}
+	id := api.SpaceID(filepath.Base(path))
+	switch conf.Storage.Kind {
+	case storage.FileStore:
+		store, err := filestore.Load(datapath)
+		if err != nil {
+			return nil, err
+		}
+		s := &fileSpace{
+			spaceBase: spaceBase{id, store, newGuard()},
+			path:      path,
+			conf:      conf,
+		}
+		return []Space{s}, nil
 
-type SearchReadCloser struct {
-	*pcap.SearchReader
-	f *os.File
-}
+	case storage.ArchiveStore:
+		store, err := archivestore.Load(datapath, conf.Storage.Archive)
+		if err != nil {
+			return nil, err
+		}
+		parent := &archiveSpace{
+			spaceBase: spaceBase{id, store, newGuard()},
+			path:      path,
+			conf:      conf,
+		}
+		ret := []Space{parent}
+		for _, subcfg := range conf.Subspaces {
+			substore, err := archivestore.Load(datapath, &storage.ArchiveConfig{
+				OpenOptions: &subcfg.OpenOptions,
+			})
+			if err != nil {
+				return nil, err
+			}
+			sub := &archiveSubspace{
+				spaceBase: spaceBase{subcfg.ID, substore, newGuard()},
+				parent:    parent,
+			}
+			ret = append(ret, sub)
+		}
+		return ret, nil
 
-func (c *SearchReadCloser) Close() error {
-	return c.f.Close()
+	default:
+		return nil, zqe.E(zqe.Invalid, "loadSpace: unknown storage kind: %s", conf.Storage.Kind)
+	}
```

Comment:
We're going to need a migration for old spaces here as they will run into this error on load- their config will have a zero value for StorageConfig and thus an Unknown value for Storage kind.

Reply:
We won't; loadConfig handles this by putting `storage.FileStore` into the Kind field if it's empty.  I also verified that I was able to create spaces with zqd on master & then load them with this PR.